### PR TITLE
[back] feat: daily job to sync youtube playlists with recommended videos

### LINKED ISSUE #1205: Push Tournesol recommendations into public YouTube playlists
Example: "Tournesol Recommendations FR": https://www.youtube.com/playlist?list=PLXX93NlSmUpZkR9NaBXMaPlPalESb5tDv

This would make Tournesol recommendations more easily accessible from mobile devices

- 2 playlists (FR and EN)
- Synchronize playlist every day to keep
  - videos published between 2 months and 3 days ago
  - with Tournesol Score >= 20

### PR DESCRIPTION
**Pull request by amatissart**
Closes #1205 

TODO in order to enable the feature on production

- [x] Create a playlist for videos in english, in addition to the existing playlist for videos in french
- [x] Get credentials associated to [Tournesol channel](https://youtube.com/@tournesolapp)
The instructions to go through the OAuth flow are described in [scripts/youtube/youtube_channel_oauth_flow.py](https://github.com/tournesol-app/tournesol/blob/yt-playlists/backend/scripts/youtube/youtube_channel_oauth_flow.py)
- [x]  Configure `YOUTUBE_CHANNEL_CREDENTIALS_JSON` on the production environment

### CODE REVIEW

**Comment by aidanjungo** [comment]
EN -> PLXX93NlSmUpayYeaNvKaeftX_QwEei3fU

**Comment by amatissart** [comment]
I've just pushed instructions about how to get the credentials associated with the Youtube channel:
https://github.com/tournesol-app/tournesol/blob/b3cd2578ab184278d6ec42ad18cb5923fad6ed0a/backend/scripts/youtube/youtube_channel_oauth_flow.py#L1-L24

This should be required only once, to add the credentials into production secrets.

File: backend/scripts/youtube/youtube_channel_oauth_flow.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+"""
+Script to generate credentials to be authenticated on Youtube API
+as the owner of a Youtube channel.
+
+Follow these instructions:
+
+1. Go to https://console.cloud.google.com/apis/credentials
+
```
Comment (aidanjungo) [inline backend/scripts/youtube/youtube_channel_oauth_flow.py:8]: Maybe we could add the first step we had to do

File: backend/tournesol/management/commands/sync_yt_playlists.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+PLAYLISTS = [
+    {"lang": "en", "playlist_id": "PLXX93NlSmUpayYeaNvKaeftX_QwEei3fU"},
+    {"lang": "fr", "playlist_id": "PLXX93NlSmUpZkR9NaBXMaPlPalESb5tDv"},
+]
+
+
+class Command(BaseCommand):
+    help = "Sync Tournesol recommendations with public playlists on Tournesol Youtube channel"
+
+    def get_videos(self, language):
```
Comment (aidanjungo) [inline backend/tournesol/management/commands/sync_yt_playlists.py:22]: Maybe we could use function name which describe a bit more what the function does. (even if a understand what it does from the return 😄)
```suggestion
    def get_top_recent_videos(self, language):
        """Get top recent video that should be present in the YouTube playlist for a certain language."""
```

File: backend/tournesol/management/commands/sync_yt_playlists.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        while request is not None:
+            response = request.execute()
+            for item in response["items"]:
+                items_dict[item["contentDetails"]["videoId"]] = item["id"]
+            request = items_client.list_next(request, response)
+
+        return items_dict
+
+    def sync_playlist(self, yt_client, playlist_id, lang):
+        videos_to_publish = {v.video_id: v for v in self.get_videos(lang)}
```
Comment (aidanjungo) [inline backend/tournesol/management/commands/sync_yt_playlists.py:49]: If the other suggestion is accepted
```suggestion
        videos_to_publish = {v.video_id: v for v in self.get_top_recent_videos(lang)}
```

File: infra/ansible/inventory.yml [hunk 4]
```diff
@@ -178,6 +178,8 @@ all:
           # different languages depending on the day of the week.
           tournesol_twitterbot_schedule: "*-*-* 18:03:00 Europe/Zurich" # daily.
 
+          tournesol_yt_playlists_schedule: "*-*-* 07:15:00" # daily
```
Comment (aidanjungo) [inline infra/ansible/inventory.yml:181]: I would put even earlier so people have fresh videos for their morning commute 😀

**Review by aidanjungo — APPROVED** [review]
Do we need some test for the playlist functions?

File: infra/ansible/inventory.yml (hunk 4 above)
Comment (amatissart) [reply infra/ansible/inventory.yml:181]: This was to be launched after the ML run, running approximately between 06:20 et 06:50 UTC

That's not very important,let's move it to 05:15 UTC

File: backend/scripts/youtube/youtube_channel_oauth_flow.py (hunk 1 above)
Comment (amatissart) [reply backend/scripts/youtube/youtube_channel_oauth_flow.py:8]: Done in c72f638e3

File: infra/ansible/inventory.yml (hunk 4 above)
Comment (amatissart) [reply infra/ansible/inventory.yml]: Done in acf3e43a

File: infra/ansible/roles/django/tasks/main.yml [hunk 5]
```diff
@@ -238,3 +238,22 @@
     state: "{{ 'started' if tournesol_twitterbot_schedule is defined else 'stopped' }}"
     enabled: "{{ 'yes' if tournesol_twitterbot_schedule is defined else 'no' }}"
     daemon_reload: yes
+
+# scheduled task: Sync Youtube playlists
+
+- name: Copy tournesol-yt-playlists service
+  template:
+    dest: /etc/systemd/system/tournesol-yt-playlists.service
+    src: tournesol-yt-playlists.service.j2
```
Comment (GresilleSiffle) [inline infra/ansible/roles/django/tasks/main.yml:247]: I haven't found this file in this PR, maybe it's not commited yet?

File: infra/ansible/roles/django/tasks/main.yml [hunk 6]
```diff
@@ -238,3 +238,22 @@
...
+
+- name: Copy tournesol-yt-playlists timer
+  template:
+    dest: /etc/systemd/system/tournesol-yt-playlists.timer
+    src: tournesol-yt-playlists.timer.j2
```
Comment (GresilleSiffle) [inline infra/ansible/roles/django/tasks/main.yml:252]: I haven't found this file too.

File: infra/ansible/roles/django/tasks/main.yml (hunk 5 above)
Comment (amatissart) [reply infra/ansible/roles/django/tasks/main.yml:247]: These 2 files had not committed. Sorry for that :clown_face:   
Fixed in cbd9a46

File: infra/ansible/roles/django/tasks/main.yml (hunk 6 above)
Comment (amatissart) [reply infra/ansible/roles/django/tasks/main.yml:252]: Fixed in cbd9a46

**Review by GresilleSiffle — APPROVED** [review]
Can't wait to see the playlists : )

Question for the future: do we also want to have a constraint on the minimal number of contributors needed by a video to be part of a playlist? Currently it's possible to reach a Tournesol score of 30 with only two contributors (both with a trusted email domain name) and six comparisons. 

For example the Twitter bots have a set of settings, maybe we would like to have the same approach, even if less strict. 
```python
# Current Twitter bots config
MIN_NB_CONTRIBUTORS = 5
MIN_NB_RATINGS = 15
MIN_RELIABILITY_SCORE = 10
MIN_TOURNESOL_SCORE = 10

# Idea of YT playlists config
MIN_NB_CONTRIBUTORS = 3 # better than 2
MIN_NB_RATINGS = MIN_NB_CONTRIBUTORS * 3 # at least 3 comparisons per user 
MIN_RELIABILITY_SCORE = 10 # to match the Twitter bots
MIN_TOURNESOL_SCORE = 20 # could be lower
```

**Comment by GresilleSiffle** [comment]
@amatissart feel free to merge and deploy when you are ready.

**Pull request merged**